### DISCUSSION TITLE
docstrings not appear in autocomplete

### DISCUSSION THREAD

**Discussion by jblindsay** [PyO3/pyo3#2846] [Questions]
Is anyone else finding that docstrings of your pyo3 project aren't displayed in VS Code/Pylance autocomplete? If I `print(my_function.__doc__)` it prints the documentation fine, but autocomplete will not display this information for me. Does anyone know why this is happening? Is it a PyO3 or a Pylance issue (my suspicions are the latter).

<img width="850" alt="Screenshot 2022-12-28 at 11 55 14 AM" src="https://user-images.githubusercontent.com/8441542/209868537-ac2c30b1-1631-4e14-a896-6b83429bfcfe.png">

**Comment by victorBR116** [comment]
Verifique se a configuração "python.languageServer" está definida como "Pylance" no seu arquivo de configuração do VS Code (settings.json). Isso garantirá que o Pylance está sendo usado como o provedor de linguagem Python para o seu código.
Em seguida, verifique se o arquivo de configuração do Pylance (pylance.json) está configurado corretamente. Você pode fazer isso verificando se as opções "python.languageServer.typeHints.preference" e "python.languageServer.completion.extraPaths" estão definidas corretamente. A opção "python.languageServer.typeHints.preference" deve estar definida como "docstring" para que as docstrings sejam exibidas no preenchimento automático. A opção "python.languageServer.completion.extraPaths" deve conter o caminho para o diretório onde o PyO3 está instalado.
Se essas configurações estiverem corretas e o problema ainda persistir, você pode tentar reiniciar o VS Code ou o provedor de linguagem Pylance. Isso pode ajudar a atualizar as configurações e garantir que as docstrings estejam sendo exibidas corretamente.

**Reply by jblindsay** [reply]
Thank you for your detailed reply. I can confirm that Pylance is set as my Python language provider. Unfortunately I do not have typeHints.preference nor a completion.extraPaths option available in my settings.

**Reply by jblindsay** [reply]
Where does one find the 'pylance.json' file? I'm assuming it can be edited along with other VS Code settings, but again, I don't see these options available there.

**Comment by victorBR116** [comment]
Para encontrar o arquivo pylance.json, basta seguir os seguintes passos:

Abra o Visual Studio Code e abra o workspace em que o PyLance foi instalado.

No menu lateral esquerdo, clique em "Explorador de Arquivos" (ou pressione Ctrl + Shift + E).

Na lista de arquivos e pastas, procure por pylance.json. Ele deve estar na raiz do seu workspace.

Por exemplo, para habilitar a análise de tipos no PyLance, você pode adicionar a seguinte configuração no arquivo pylance.json:

{
  "python.languageServer": "Pylance",
  "python.analysis.typeCheckingMode": "basic"
}

desculpas pela demora na resposta.

**Reply by jblindsay** [reply]
I think it might be a Mac thing because I have no pylance.json file in the root of my workspace unfortunately (and I checked in the .vscode folder as well). In fact, there is no such file anywhere on my machine by the looks of it. Pylance is definitely being used by VSCode however; I have confirmed that. It's really just this one aspect of not showing the help docs of my PyO3 project. No worries about the slow response, particularly given the time of year. Thanks again for your help.

**Comment by cgevans** [comment]
It appears that Pylance cannot read docstrings in non-Python code, as it does not execute code, and is instead statically analyzing files.  For major projects (eg, pandas), it gets around this by having stub files with the docstrings there, so it seems like it works. 

Unfortunately, this means docstrings in Rust code show up in Python, but not via Pylance, and docstrings in stub files show up in VS Code, but not in Python.  Some way of synchronizing the docstrings in the two seems like it would be the only effective option right now.

See https://github.com/microsoft/pylance-release/issues/170, https://github.com/microsoft/pylance-release/discussions/2711, or https://github.com/microsoft/pylance-release/issues/49.  There appears to be some limited support, but only for actual builtins, and that feature appears to be a closed-source value-added feature for pylance (see https://github.com/microsoft/pyright/issues/4250#issuecomment-1331647419).

**Reply by MusicalNinjaDad** [reply]
Hey @cgevans, yes - microsoft have created stubs for the most popular libraries which are lacking them: pylance would be "unusable" without hinting for e.g. pandas. Other libraries have to create, maintain & distribute the stubs themselves, the "power dynamic" is a bit different ;)

I put `pyo3-stubgen` together for this purpose. It's full functional but bare-bones right now but I intend to continue to maintain it both as a CLI app and via a library API. [pypi](https://pypi.org/project/pyo3-stubgen/); [docs](https://musicalninjas.github.io/pyo3-stubgen/); [github](https://github.com/MusicalNinjas/pyo3-stubgen)

A bit more background and usage ideas are here: [FizzBuzz - Combining rust & python](https://musicalninjadad.github.io/FizzBuzz/dev-build/#ide-type-doc-hinting)

**Reply by cgevans** [reply]
Yes! Thank you for putting that together.  With the still-incomplete modifications for class and method support and numpydoc type parsing I have draft PRs for, I have been able to start using it to put together stub files with minimal manual editing.

**Reply by MusicalNinjaDad** [reply]
Thanks - I hadn't spotted those PRs, I've just taken a quick look and I think they add really valuable functionality!

**Comment by toppk** [comment]
I just created a .pyi file.  Sure it's redundant typing information, but it allows you to put in pydocs that will just work with all language servers out of the box. .
